### PR TITLE
Connect GUI to GraphPanel + minor improvements

### PR DESCRIPTION
**Pull request by FabianRolfMatthiasNoll**

### CODE REVIEW

File: src/GUI/GraphPanel.java [hunk 1]
```diff
@@ -165,7 +175,25 @@ private void drawGrid(Graphics2D g2d, int width, int height, double step) {
...
+        path.moveTo(startPosition.x, startPosition.y);
+
+        for(int i = 1; i <= expression.numSamples; i++) {
+            double t = expression.lowerBound + i * sampleLength;
+            Vector2D position = toScreenCoordinates(expression.evaluate(t));
+            path.lineTo(position.x, position.y);
+        }
+
+        g2d.setStroke(new BasicStroke(2.0f));
+        g2d.setColor(Color.WHITE);
```
Comment (FabianRolfMatthiasNoll) [inline src/GUI/GraphPanel.java]: If there is now a drawParametricFunction differentiating in the drawPolynomialFunctions is not needed anymore. (Will fix that in a sec)

Comment (FabianRolfMatthiasNoll) [reply src/GUI/GraphPanel.java]: Done

File: README
Comment (kantholz44mm) [inline README:1]: y tho

File: src/GUI/ComboBoxColorPicker.java [hunk 2]
```diff
@@ -1,42 +1,43 @@
...
 import javax.swing.*;
 import java.awt.*;
 import java.awt.event.ActionEvent;
 import java.awt.event.ActionListener;
 
 class ComboBoxColorPicker extends JPanel {
     private final Color[] graphColors = {Color.RED, Color.BLUE, Color.YELLOW, Color.GREEN, Color.GRAY};
-    private final JComboBox<Color> colorPicker = new JComboBox<>(graphColors);   // Combo Box to pick the Color of the Function
+    private final JComboBox<Color> colorPicker = new JComboBox<>(graphColors);
+    public Color currentColor;
```
Comment (kantholz44mm) [inline src/GUI/ComboBoxColorPicker.java]: avoid state where possible. Rather, let's have a function called 'getCurrentColor' (or similar) and just return

`(Color) colorPicker.getSelectedItem()`

from there. 'currentColor' is not needed as far as I can see.

File: src/GUI/ComboBoxColorPicker.java [hunk 3]
```diff
@@ -1,42 +1,43 @@
...
+        currentColor = (Color) colorPicker.getSelectedItem();
         add(colorPicker, BorderLayout.CENTER);
     }
 
     private class ColorBoxListener implements ActionListener {
         @Override
         public void actionPerformed(ActionEvent e) {
-            Color selectedColor = (Color) colorPicker.getSelectedItem(); // Gets the selected Color from the ComboBox (cast to Color)
-            //TODO: Connect to the color setter of the function
+            currentColor = (Color) colorPicker.getSelectedItem();
```
Comment (kantholz44mm) [inline src/GUI/ComboBoxColorPicker.java]: Without currentColor, this whole actionListener can be omitted.

File: src/GUI/FunctionControlPanel.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+public class FunctionControlPanel extends JPanel {
+    private final GraphPanel graphPanel;
+    private int instanceID;
+    private final GridBagLayout InputLayout = new GridBagLayout();
+    private final GridBagConstraints gbc = new GridBagConstraints();
+    private final List<FunctionControlPanel> functionControlPanelList;
+    private final JPanel functionInputPosition;
+    private final PolynomialFunctionControl polynomialFunctionControl;
+    private ComboBoxColorPicker colorPicker = null;
+    JTextField functionField = new JTextField(20);
```
Comment (kantholz44mm) [inline src/GUI/FunctionControlPanel.java]: missing visibility specifier

File: src/GUI/GraphPanel.java [hunk 5]
```diff
@@ -45,77 +51,67 @@ public void mouseMoved(MouseEvent e){
...
-                PolynomialFunction function = new PolynomialFunction(functionString);
-                function.calcRoots(minT, maxT, 0.001);
-                function.calcExtremePoints(minT, maxT, 0.001);
-                functions.add(function);
-                repaint();
-            } else {
-                GraphPanel.infoBox("You reached the maximum amount of Graphs", "MAX_GRAPHS_REACHED");
-            }
-
+    public int addFunction(String functionString, Color currentColor){
```
Comment (kantholz44mm) [inline src/GUI/GraphPanel.java:58]: maybe a comment is warranted here for documenting what is returned. Just from the name, I wouldn't know what the integer, returned by 'addFunction' means.

File: src/GUI/GraphPanel.java [hunk 6]
```diff
@@ -180,31 +175,46 @@ private void drawGrid(Graphics2D g2d, int width, int height, double step) {
...
+            Vector2D position = toScreenCoordinates(expression.evaluate(t));
+            path.lineTo(position.x, position.y);
+        }
 
+        g2d.setStroke(new BasicStroke(2.0f));
+        g2d.setColor(Color.WHITE);
+        g2d.draw(path);
+    }
+
+    private void drawPolynomialFunctions(Graphics2D g2d, int width) {
```
Comment (kantholz44mm) [inline src/GUI/GraphPanel.java]: now that I think about it, we should just have one generic 'drawParametricFunction' that takes as input a parametric function, start and end range, steps and Color and then just draws it. The only difference between Polynomials and Parametrics so far is that polynomials get their start and end coords from the screen edges. That shouldn't happen in the 'draw' function, though

**Review by kantholz44mm — CHANGES_REQUESTED** [review]
ye looks good, some minor nitpicks, but otherwise gud 2 go

File: README
Comment (FabianRolfMatthiasNoll) [reply README:1]: wtf where does this change come from???

File: src/GUI/ComboBoxColorPicker.java (hunk 2 above)
Comment (FabianRolfMatthiasNoll) [reply src/GUI/ComboBoxColorPicker.java]: Done

File: src/GUI/ComboBoxColorPicker.java (hunk 3 above)
Comment (FabianRolfMatthiasNoll) [reply src/GUI/ComboBoxColorPicker.java]: Done

File: src/GUI/FunctionControlPanel.java (hunk 4 above)
Comment (FabianRolfMatthiasNoll) [reply src/GUI/FunctionControlPanel.java]: Done

File: src/GUI/GraphPanel.java (hunk 5 above)
Comment (FabianRolfMatthiasNoll) [reply src/GUI/GraphPanel.java:58]: Done

File: src/GUI/GraphPanel.java (hunk 6 above)
Comment (kantholz44mm) [reply src/GUI/GraphPanel.java]: done

**Review by kantholz44mm — APPROVED** [review]

**Pull request merged**